src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -250,231 +220,94 @@ public void AddPeer(ITxPoolPeer peer)
+        private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast)
         {
-            bool isInHashCache = _hashCache.Get(tx.Hash);
-            // !!! do not change it to |=
-            bool isKnown = !isReorg && isInHashCache;
-
-            if (!isInHashCache)
-            {
-                _hashCache.Set(tx.Hash);
-            }
-
-            /*
-             * we need to make sure that the sender is resolved before adding to the distinct tx pool
-             * as the address is used in the distinct value calculation
-             */
-            if (!isKnown)
+            lock (_locker)
             {
-                lock (_locker)
+                bool eip1559Enabled = _specProvider.GetSpec().IsEip1559Enabled;
+                
+                // TODO: is it called Bottleneck or effective gas price?
+                tx.GasBottleneck = tx.CalculateEffectiveGasPrice(eip1559Enabled, _headInfo.CurrentBaseFee);
+                bool inserted = _transactions.TryInsert(tx.Hash, tx, out Transaction? removed);
+                if (inserted)
                 {
-                    tx.GasBottleneck = tx.CalculateEffectiveGasPrice(_specProvider.GetSpec().IsEip1559Enabled, CurrentBaseFee);
-                    bool inserted = _transactions.TryInsert(tx.Hash, tx, out Transaction? removed);
-                    if (inserted)
-                    {
-                        _transactions.UpdateGroup(tx.SenderAddress, UpdateBucketWithAddedTransaction);
-                        ForgetHashOfRemovedTransaction(removed?.Hash);
-                        Metrics.PendingTransactionsAdded++;
-                        if (tx.IsEip1559)
-                        {
-                            Metrics.Pending1559TransactionsAdded++;
-                        }
-                    }
-                    isKnown |= !inserted;
+                    _transactions.UpdateGroup(tx.SenderAddress!, UpdateBucketWithAddedTransaction);
+                    Metrics.PendingTransactionsAdded++;
+                    if (tx.IsEip1559) { Metrics.Pending1559TransactionsAdded++; }
+
+                    // transaction which was on last position in sorted TxPool and was deleted to give
+                    // a place for a newly added tx (with higher priority) is now removed from hashCache
+                    // to give it opportunity to come back to TxPool in the future, when fees drops
+                    Metrics.PendingTransactionsEvicted++;
+                    if(removed != null) _hashCache.Delete(removed.Hash!);
+                }
+                else
+                {
+                    return AddTxResult.FeeTooLowToCompete;
                 }
             }
 
-            if (!isKnown)
-            {
-                isKnown |= !isReorg && (_txStorage.Get(tx.Hash) is not null);
-            }
-
-            if (isKnown)
+            // TODO: this is strange as it may pretend that it has skipped the transaction while the transaction has been added above?
```

Comment:
Do we need _txStorage at all here? It is doing IO and using most resources here

Reply:
_txStorage is no longer useful at this point. I think we should remove it.

Reply:
I was thinking that it may be the case. I suggets removing after you confirm but please remove as a separate PR